scapy/contrib/automotive/xcp/xcp.py

```diff
@@ -0,0 +1,422 @@
+from scapy.packet import Packet, bind_layers, bind_bottom_up
+
+if 'XCP' not in conf.contribs:
+    conf.contribs['XCP'] = {}
+
+# 0 stands for Intel/little-endian format, 1 for Motorola/big-endian format
+conf.contribs['XCP']['byte_order'] = 1
+conf.contribs['XCP']['allow_byte_order_change'] = True
+conf.contribs['XCP']['Address_Granularity_Byte'] = None  # Can be 1, 2 or 4
+conf.contribs['XCP']['allow_ag_change'] = True
+
+conf.contribs['XCP']['MAX_CTO'] = None
+conf.contribs['XCP']['MAX_DTO'] = None
+conf.contribs['XCP']['allow_cto_and_dto_change'] = True
+
+
+# Specifications from:
+# http://read.pudn.com/downloads293/doc/comm/1316424/ASAM_XCP_Part1-Overview_V1.0.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%202-%20Protocol%20Layer%20Specification%20-1.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%203-%20Transport_layer_specification_xcp_on_can_1-0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%204-%20Interface%20Specification%20-1.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%205-%20Example%20Communication%20Sequences%20-1.0.pdf # noqa: E501
+
+# XCP on USB is left out because it has "no practical meaning"
+# XCP on Lin is left out because it has no official specification
+class XCPOnCAN(CAN):
+    name = 'Universal calibration and measurement protocol on CAN'
+    fields_desc = [
+        FlagsField('flags', 0, 3, ['error',
+                                   'remote_transmission_request',
+                                   'extended']),
+        XBitField('identifier', 0, 29),
+        ByteField('length', 8),
+        ThreeBytesField('reserved', 0),
+    ]
+
+    def extract_padding(self, p):
+        return p, None
+
+
+class XCPOnUDP(UDP):
+    name = 'Universal calibration and measurement protocol on Ethernet'
+    fields_desc = [
+        ShortEnumField("sport", 53, UDP_SERVICES),
+        ShortEnumField("dport", 53, UDP_SERVICES),
+        ShortField("len", None),
+        XShortField("chksum", None),
+        #  XCP Header
+        ShortField('length', None),
+        ShortField('ctr', 0),  # counter
+    ]
+
+
+class XCPOnTCP(TCP):
+    name = 'Universal calibration and measurement protocol on Ethernet'
+
+    fields_desc = [
+        ShortEnumField("sport", 20, TCP_SERVICES),
+        ShortEnumField("dport", 80, TCP_SERVICES),
+        IntField("seq", 0),
+        IntField("ack", 0),
+        BitField("dataofs", None, 4),
+        BitField("reserved", 0, 3),
+        FlagsField("flags", 0x2, 9, "FSRPAUECN"),
+        ShortField("window", 8192),
+        XShortField("chksum", None),
+        ShortField("urgptr", 0),
+        TCPOptionsField("options", ""),
+        #  XCP Header
+        ShortField('length', None),
+        ShortField('ctr', 0),  # counter
+    ]
+
+    def answers(self, other):
+        if other.__class__ != self.__class__:
```

Comment:
I guess the more pythonic way would be
`if type(other) != type(self):`

See https://stackoverflow.com/a/10633356/2050020